lib/stages/nDiskMultiFormatWriter.cpp

```diff
@@ -0,0 +1,392 @@
+    }
+
+    // Join the threads
+    for (uint32_t i = 0; i < num_disks; ++i) {
+        file_thread_handles[i].join();
+    }
+}
+
+/// writes the incoming frames to a raw file
+void nDiskMultiFormatWriter::raw_file_write_thread(int disk_id) {
+    int fd = -1;
+    size_t file_num = disk_id;
+    int frame_id = disk_id;
+    uint8_t* frame = nullptr;
+    int64_t num_frames = 0;
+    std::string file_name;
+    ssize_t bytes_written;
+
+    // thread infinite loop
+    while (!stop_thread) {
+
+        // This call is blocking.
+        frame = wait_for_full_frame(buf, unique_name.c_str(), frame_id);
+        if (frame == nullptr)
+            break;
+
+        // INFO("Got buffer id: {:d}, disk id {:d}", frame_id, disk_id);
+
+        // Check if the producer has finished, and we should exit.
+        if (frame_id == -1) {
+            break;
+        }
+
+        // Open the file to write
+        if (write_to_disk) {
+
+            if (fd < 0) {
+                // make the output file name
+                file_name = get_dataset_folder_name(disk_id) + fmt::sprintf("/%010zu.", file_num)
+                            + file_extension;
+                // open a file and get its handle
+                fd = open(file_name.c_str(), O_WRONLY | O_CREAT, 0666);
+
+                // set num_frames to -1 just to write -1 at the beginning of the file
+                num_frames = -1;
+                // write the number of frames in this file
+                bytes_written = write(fd, &num_frames, sizeof(num_frames));
+                // reset the num_frames
+                num_frames = 0;
+            }
+
+            if (fd == -1) {
+                ERROR("Cannot open file");
+                ERROR("File name was: {:s}", file_name);
+                exit(errno);
+            }
+
+            // write the metadata and its size
+            size_t metadata_size = 0;
+            struct metadataContainer* mc = get_metadata_container(buf, frame_id);
+            if (mc != nullptr) {
+                metadata_size = mc->metadata_size;
+            }
+            // Write metadata size to disk, if there is no metadata in the frame, then
+            // just save 0 to the first word.
+            if (write(fd, (void*)&metadata_size, sizeof(metadata_size))
+                != (int32_t)sizeof(metadata_size)) {
+                ERROR("Failed to write metadata_size to disk for file {:s}", file_name);
+                exit(-1);
+            }
+            if (mc != nullptr) {
+                if (write(fd, mc->metadata, mc->metadata_size) != (int32_t)mc->metadata_size) {
+                    ERROR("Failed to write metadata_size to disk for file {:s}", file_name);
+                    exit(-1);
+                }
+            }
+
+            // write the number of frequencies
+            bytes_written =
+                write(fd, &num_freq_per_output_frame, sizeof(num_freq_per_output_frame));
```

Comment:
I think this `num_freq_per_output_frame` should be written into the output frame in the previous step (or put into the metadata).   If you do that, then this code just writes the frame size and frame without needing to know the type of data it is writing, which makes it a lot more flexible.

Reply:
It is define as an input config parameter for BeamAssemble stage. If I add it as the leading uint64_t in each frame, we need to change the buffer size by adding 4 as well, right?